toolchain/check/handle_modifier.cpp

```diff
@@ -0,0 +1,116 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/context.h"
+#include "toolchain/lex/token_kind.h"
+
+namespace Carbon::Check {
+
+CARBON_DIAGNOSTIC(ModifierNotAllowedWith, Error,
+                  "`{0}` not allowed on declaration with `{1}`.",
+                  llvm::StringRef, llvm::StringRef);
+CARBON_DIAGNOSTIC(ModifierRepeated, Error, "`{0}` repeated on declaration.",
+                  llvm::StringRef);
+CARBON_DIAGNOSTIC(ModifierMustAppearBefore, Error,
+                  "`{0}` must appear before `{1}`.", llvm::StringRef,
+                  llvm::StringRef);
+CARBON_DIAGNOSTIC(ModifierPrevious, Note, "`{0}` previously appeared here.",
+                  llvm::StringRef);
+
+static auto AccessModifierEnum(Lex::TokenKind token_kind) -> auto {
```

Comment:
Why are you using an `auto` return here, versus specifying the type? I'm not clear this makes the code safer, [per style](https://google.github.io/styleguide/cppguide.html#Type_deduction).